Test/PublishNodeModule.Tests.ps1

```diff
@@ -156,6 +158,41 @@ function ThenPackageJsonVersion
     $packageJson | Select-Object -ExpandProperty 'version' | Should -Be $ExpectedVersion
 }
 
+function ThenPublishedWithTag
+{
+    param(
+        $ExpectedTag
+    )
+
+    Assert-MockCalled   -CommandName 'Invoke-WhiskeyNpmCommand' `
+                        -ModuleName 'Whiskey' `
+                        -Times 1 `
+                        -Exactly `
+                        -ParameterFilter {
+                            if( $Name -ne 'publish')
+                            {
+                                return $false

```

Comment:
I've never liked how hard it is to tell what parameter was bad in these filters. Last week, I had the idea to add `Write-Error` statements to these filters so I didn't have to write and enable debug statements. I was going to suggest that, but then I thought "Why not just add `Should` assertions?`

Idid some testing and it looks like you *can* use `Should` assertions inside  parameter filters. I think we should move to a pattern where instead of returning `$false`, we make an assertion. Then, after all the assertions, return `$true`.

```powershell
$Name | Should -Be 'publish' -Because 'Invoke-NpmCommand should be called with publish command.'
'--tag' | Should -BeIn $ArgumentList -Because 'Invoke-NpmCommand should be called with "--tag" argument'
$ExpectedTag | Should -BeIn $ArgumentList -Because 'Invoke-NpmCommand should be called with expected tag argument'
$ArgumentList | Should -HaveCount 2 'Invoke-NpmCommand shouldn''t be called with extra arguments'
return $true
```
```

Reply:
This does work well, though only if there is a single call to the mocked command with the expected parameter values. When there are multiple calls to the mocked command with different parameter values, you must mix the two patterns, which I personally don't mind:

```
-ParameterFilter {
    if( $Name -ne 'publish')
    {
        return $false
    }

    '--tag' | Should -BeIn $ArgumentList -Because 'Invoke-NpmCommand should be called with "--tag" argument'
    $ExpectedTag | Should -BeIn $ArgumentList -Because 'Invoke-NpmCommand should be called with expected tag argument'
    $ArgumentList | Should -HaveCount 2 -Because 'Invoke-NpmCommand shouldn''t be called with extra arguments'

    return $true
}
```